### PR TITLE
test/e2e: Pull deployment manifest unmarshal/update into framework

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Pulls potentially commonly used bits into e2e framework for use in other
suites
- Deployment automatically unmarshalled when e2e framework created
- Pattern generally is: modify resources as test needs from the base
manifest, deploy, maybe wait for some conditions, test
- Method to ensure each piece of deployment, will upsert resource

First part of a series that should start us on the path to running contour locally in e2e tests, other commits on the way, this is the base and self contained so felt like it was good to PR before other things get finished

Updates: #3772

### CODE REVIEW

File: test/e2e/deployment.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	EnvoyDaemonSet            *apps_v1.DaemonSet
+}
+
+// Unmarshals resources from rendered Contour manifest in order.
+// Note: This will need to be updated if any new resources are added to the
+// rendered deployment manifest.
+func (d *Deployment) UnmarshalResources() error {
+	_, thisFile, _, ok := runtime.Caller(0)
+	if !ok {
+		return errors.New("could not get filename")
```
Comment (stevesloka) [inline test/e2e/deployment.go:65]: nit: Add the filename to the error message?

File: test/e2e/upgrade/upgrade_test.go [hunk 2]
```diff
@@ -178,201 +137,78 @@ func checkRoutability(host string) {
...
-
-	return resources
+	require.Len(f.T(), f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers, 1)
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers[0].Image = contourUpgradeToImage
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.InitContainers[0].ImagePullPolicy = v1.PullIfNotPresent
+	require.Len(f.T(), f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers, 2)
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers[0].Image = contourUpgradeToImage
+	f.Deployment.EnvoyDaemonSet.Spec.Template.Spec.Containers[0].ImagePullPolicy = v1.PullIfNotPresent
+	require.NoError(f.T(), f.Deployment.EnsureEnvoyDaemonSet())
+	// TODO: when we deploy the from version in this test, ensure envoy
```
Comment (stevesloka) [inline test/e2e/upgrade/upgrade_test.go:212]: Does this need an issue to track?

File: test/e2e/deployment.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	CertgenJob                *batch_v1.Job
+	ContourClusterRoleBinding *rbac_v1.ClusterRoleBinding
+	ContourClusterRole        *rbac_v1.ClusterRole
+	ContourService            *v1.Service
+	EnvoyService              *v1.Service
+	ContourDeployment         *apps_v1.Deployment
+	EnvoyDaemonSet            *apps_v1.DaemonSet
+}
+
+// Unmarshals resources from rendered Contour manifest in order.
```
Comment (skriss) [inline test/e2e/deployment.go:59]:
```suggestion
// UnmarshalResources unmarshals resources from rendered Contour manifest in order.
```

File: test/e2e/deployment.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+func (d *Deployment) EnsureContourDeployment() error {
+	return d.ensureResource(d.ContourDeployment, new(apps_v1.Deployment))
+}
+
+func (d *Deployment) WaitForContourDeploymentUpdated() error {
+	// List pods with app label "contour" and check that pods are updated
+	// with expected container image and in ready state.
+	// We do this instead of checking Deployment status as it is possible
+	// for it not to have been updated yet and replicas not yet been shut
+	// down.
```
Comment (skriss) [inline test/e2e/deployment.go:219]: 👍  nice solution and comment.

**Review by skriss — COMMENTED** [review]
Looks pretty good to me, just one nit

File: test/e2e/upgrade/upgrade_test.go (hunk 2 above)
Comment (sunjayBhatia) [reply test/e2e/upgrade/upgrade_test.go:212]: I'll remove TODOs and put them in https://github.com/projectcontour/contour/issues/3679 instead

**Review by skriss — APPROVED** [review]

**Review by stevesloka — APPROVED** [review]

**Pull request merged**